### ISSUE TITLE
random_bytes() hangs at 100% CPU

### ISSUE THREAD

**Issue by xTeixeira** [paragonie/random_compat#145]
Hello.

I have a cron job that runs a PHP script every minute, which at some point calls random_bytes().
Sometimes this works normally (the process runs for ~2 seconds and exits with success), but sometimes the PHP process hangs at 100% CPU and never dies unless I kill it manually.

Using xdebug to get a strack trace and setting a maximum execution time I found the point it hangs at:
`PHP Fatal error: Maximum execution time of 20 seconds exceeded in /var/www/vendor/paragonie/random_compat/lib/random_bytes_dev_urandom.php on line 144`

I also noticed this was similar to #90, but I'm not sure if it's really related.

**Comment by paragonie-scott** [comment]
It's not #90, because `random_bytes_dev_urandom.php` doesn't use mcrypt.

What OS? What PHP version? Are you in any special envrionment (FreeBSD jail, for example)?

**Comment by xTeixeira** [comment]
I'm on CentOS 7.5, PHP 5.6.36, random_compat 2.0.15.

Only "special" thing about the environment is that it's a VM.

**Comment by paragonie-scott** [comment]
Can you give `dev-master` a try?

**Comment by xTeixeira** [comment]
Tried it, unfortunately it didn't work, still occasionally hangs in the same way as before.

**Comment by paragonie-scott** [comment]
How many random bytes are you generating? How much RAM is on the server?

I'm not sure what the problem could be, but I'm determined to find out. Did the line number change for the timeout?

**Comment by xTeixeira** [comment]
Last few times I tested the line number didn't change.

I tried removing the random_bytes() from the script and the process stopped hanging (even without any timeout), so it definitely hangs on random_bytes(), although I definitely think it may be possible that another part of the code somehow messes up the memory and that causes random_bytes() to fail and hang, right? I've been trying to find other possible culprits but found nothing so far.

One additional info, I ran the script with `strace` and found that when it hangs it keeps running a `read()` syscall over and over (that's why it goes to 100% CPU usage I guess) and it kept failing with a bad file descriptor error. I don't have the `strace` output right now but I do remember the file descriptor used for `read()` was always 0.

Any thoughts on that?

**Comment by paragonie-scott** [comment]
Is `/dev/urandom` readable from PHP?

**Comment by xTeixeira** [comment]
I would think so, my `/dev/urandom` permissions are 666. Plus the script works normally sometimes so I'd guess in those times it reads `/dev/urandom` with no problem, right?

**Comment by paragonie-scott** [comment]
This might be a PHP bug. We *are* caching the file handler throughout the lifetime of the script.

Pull `dev-master`, remove the `static` from `static $fp = null;` on line 51. Does the problem go away?

**Comment by xTeixeira** [comment]
Hey, sorry for the delay in answering.
I've just now confirmed that the issue was not in random_compat, but another library that created memory errors and the program would crash later in random_bytes().
I've successfully recreated the error without using random_compat and observed that the program would crash anyway in some cases. Removing the use of the culprit library solves the issue completely.

Sorry for the confusion and thanks for the help in narrowing down the issue!
I'm closing this now.
